# Kinema 

<br>

## A Film recommendation system based on below the line department heads

<br>

**These departments and position include**:

- **Writing** - Writer, Publisher, Screenplay, Story By...
- **Camera** - Director of Photography
- **Electric** - Lighting
- **Grip** - Stunts, Rigging, SFX
- **Art** - Production Designer, Prop Master, Art Director, Scenic
- **GLAM** - Hair, Make Up, Wardrobe
- **Editing** - Editor(s)
- **Sound** - Scorer, Music Supervisor, Foley Artist
- **Color** - Colorist, SFX, Lighting Technician, 
- **Locations** - Location Manager

As someone who has been passionately Directing, Writing, Acting, and Working professionaly in the Film industry for the past 8 years in Los Angeles, CA there is one thing (it's actually many!) that must be said when it comes to the streaming of Film and Television. The standard film studios use to recommendation films (recommender systems) needs to be reimagined and implemented.
<br>

That's where the unsung heroes of film magic come in. The professionals, the dreamers, the artists that make films come together are overlooked and under represented within the film industry. Most film recommendation systems follow a Film Director, Actor, and Movie Rating approach while also splashing in some random films to give user's that "organic" experience giving them the facade of natural selection. There is something critically missing from this entire approach. 
<br>

***What makes the film?***
<br>

> *It is not what we see, it is what we feel.*


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import norm
from scipy import stats

In [2]:
credits_df = pd.read_csv('data/credits.csv')
keywords_df = pd.read_csv('data/keywords.csv')
links_small_df = pd.read_csv('data/links_small.csv')
links_df = pd.read_csv('data/links.csv')
movies_metadata_df = pd.read_csv('data/movies_metadata.csv')
ratings_df = pd.read_csv('data/ratings.csv')

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
movies_metadata_df

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisati

In [4]:
movie_df = pd.read_csv('data/movies_metadata.csv')
movie_df.drop_duplicates(subset=['title'], inplace=True)
movie_df = movie_df.reset_index().drop('index', axis=1)
movie_df.head()

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [5]:
# Create dataframe with certain features
movie_df = movie_df[['id', 'title', 'overview', 'vote_average', 'vote_count']]
movie_df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  vote_average  vote_count  
0  Led by Woody, Andy's toys live happily in his ...           7.7      5415.0  
1  When siblings Judy and Peter discover an encha...           6.9      2413.0  
2  A family wedding reignites the ancient feud be...           6.5        92.0  
3  Cheated on, mistreated and stepped on, the wom...           6.1        34.0  
4  Just when George Banks has recovered from his ...           5.7       173.0

In [6]:
movie_df.shape

(42278, 5)

### A User Independent System

We'll start with building recommendations for the case where we have no knowledge about a user. In this case, we probably would find the most popular items, and recommend them to our user.

The simplest system would just sort the items based on their average score and present a list in that order:

In [7]:
# Sort movies by rating
top_rated = movie_df.sort_values(by=['vote_average'], ascending=False)
top_rated['title'].head(10)

40390                          Jedi Junior High
30753                         OMG, I'm a Robot!
36536                             Truckfighters
34225                                    Avetik
28951         Mortal Kombat: The Journey Begins
8704     Common Threads: Stories from the Quilt
29183                                 Time Pass
29339                   The Mother Of Invention
34224                                  Acéphale
29771                       The Kreutzer Sonata
Name: title, dtype: object

In [8]:
top_rated.head(10)

id                                   title  \
40390  426224                        Jedi Junior High   
30753  347328                       OMG, I'm a Robot!   
36536  206657                           Truckfighters   
34225  237422                                  Avetik   
28951   72886       Mortal Kombat: The Journey Begins   
8704   114524  Common Threads: Stories from the Quilt   
29183  249677                               Time Pass   
29339  140489                 The Mother Of Invention   
34224  243416                                Acéphale   
29771   53315                     The Kreutzer Sonata   

                                                overview  vote_average  \
40390  In 2009, students of all walks of life at the ...          10.0   
30753         A sensitive guy finds out he's... a robot.          10.0   
36536  Three down-to-earth guys, record songs and rea...          10.0   
34225  Hovering between the realms of poetry and hist...          10.0   
28951  The heroes from the hard-hitting MORTAL KOMBAT...          10.0   
8704   In the late 1970s, a mysterious new disease be...          10.0   
29183  When Cupid strikes the very first time it sees...          10.0   
29339  A mockumentary about an aspiring inventor who ...          10.0   
34224  In common with other Zanzibar films, Patrick D...          10.0   
29771  A film by Bernard Rose, based on the novel by ...          10.0   

       vote_count  
40390         1.0  
30753         1.0  
36536         1.0  
34225         3.0  
28951         1.0  
8704          2.0  
29183         1.0  
29339         1.0  
34224         1.0  
29771         1.0

I can notice one large issue with the average score within this system. The majority of perfect average scores have very low vote counts and should not be included when being compared with films that have 100+ reviews(`vote_count`)

Take a look at a popular film and understand how many votes that film has and what the average rating is for that film to attain a better understanding 'realistically' of how many reviews a film should have and what the average rating would be.

<br>

The movie that quickly comes to mind for me is ["Ex Machina"](https://www.imdb.com/title/tt0470752/)

- let's bring up the information for that particular film.

In [10]:
# Retrieve film Ex machina
top_rated[top_rated['title']== "Ex Machina"]

id       title                                           overview  \
23191  264660  Ex Machina  Caleb, a 26 year old coder at the world's larg...   

       vote_average  vote_count  
23191           7.6      4862.0

Ex Machina is a widely popular movie and as we can see has a total vote_count of 4,862 and this number is more indicative of the number of reviews a film should be recieving.
<br>

I will deal with the films with minimal reviews by creating a threshold in which we ignore items with too few reviews.

In [11]:
# Initiate threshold count
threshold = 1000
# Retrieve movies that break the threshold
movies_with_many_reviews_df = movie_df[movie_df['vote_count'] >= threshold]

# Create new dataframe for movies above 1k votes
top_rated_v2 = movies_with_many_reviews_df.sort_values(by=['vote_average'], ascending=False)
top_rated_v2.head(10)

id                            title  \
834       238                    The Godfather   
314       278         The Shawshank Redemption   
37533  372058                       Your Name.   
1150      510  One Flew Over the Cuckoo's Nest   
2814      550                       Fight Club   
1174      539                           Psycho   
1176      240           The Godfather: Part II   
1182      311      Once Upon a Time in America   
292       680                     Pulp Fiction   
522       424                 Schindler's List   

                                                overview  vote_average  \
834    Spanning the years 1945 to 1955, a chronicle o...           8.5   
314    Framed in the 1940s for the double murder of h...           8.5   
37533  High schoolers Mitsuha and Taki are complete s...           8.5   
1150   While serving time for insanity at a state men...           8.3   
2814   A ticking-time-bomb insomniac and a slippery s...           8.3   
1174   When larcenous real estate clerk Marion Crane ...           8.3   
1176   In the continuing saga of the Corleone crime f...           8.3   
1182   A former Prohibition-era Jewish gangster retur...           8.3   
292    A burger-loving hit man, his philosophical par...           8.3   
522    The true story of how businessman Oskar Schind...           8.3   

       vote_count  
834        6024.0  
314        8358.0  
37533      1030.0  
1150       3001.0  
2814       9678.0  
1174       2405.0  
1176       3418.0  
1182       1104.0  
292        8670.0  
522        4436.0

In [12]:
df_descriptions = movie_df[['title', 'overview']]
df_descriptions.head(10)

title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   
5                         Heat   
6                      Sabrina   
7                 Tom and Huck   
8                 Sudden Death   
9                    GoldenEye   

                                            overview  
0  Led by Woody, Andy's toys live happily in his ...  
1  When siblings Judy and Peter discover an encha...  
2  A family wedding reignites the ancient feud be...  
3  Cheated on, mistreated and stepped on, the wom...  
4  Just when George Banks has recovered from his ...  
5  Obsessive master thief, Neil McCauley leads a ...  
6  An ugly duckling having undergone a remarkable...  
7  A mischievous young boy, Tom Sawyer, witnesses...  
8  International action superstar Jean Claude Van...  
9  James Bond must unmask the mysterious head of ...

In [13]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(stop_words = "english", min_df=2)
movie_df['overview'] = movie_df['overview'].fillna("")

TF_IDF_matrix = vectorizer.fit_transform(movie_df['overview'])

In [14]:
# 42,728 movies by 36,504 tokens in the descriptions =~ 1.5B elements!
TF_IDF_matrix.shape

(42278, 36504)